### ISSUE TITLE
Parallel handling of LSP requests/notifications and cancellation

### ISSUE THREAD

**Issue by cocreature** [digital-asset/daml#1741]
Currently we have a single event queue that handles requests send my the LSP client and processes them one my one. This is suboptimal in two ways:

1. Some requests can be run in parallel, e.g., a request for codelenses and a request for hover.
2. If we get a notification that the document has changed, we keep processing all requests in the queue even if they are already outdated until we get to the change notification in the queue.

LSP gives us some liberty here but also makes some requirements:

1. All requests need to be answered. It is possible to answer with an error response but not responding at all is invalid.
2. It is fine to reorder and parallelize requests but only if it does not change the result. So if we get something like a document change request, we are not allowed to process that first and then respond based on the updated document.
3. LSP supports explicit request cancellation. 
> if a client sends a request to the server and the client state changes in a way that the result will be invalid it should cancel the server request and ignore the result.

Based on this, I think the benefits we gain from parallelizing requests are actually fairly small since most requests (with the notable exceptions of codelenses) are triggered by the user there is rarely more than one request running and even if there are mutiple requests they are probably waiting on the same thing, e.g., typechecking of the currently opened file so running them in parallel doesn’t buy us much. The fact that our Shake API doesn’t have an easy way of adding a new request without cancelling the current one also makes this somewhat tricky to do (we could cancel the existing one and then restart both but if we almost finished an expensive typecheck, this might actually make things worse).

What does seem useful is to cancel requests if contents are updated. We have two options for that:

1. We look on the server side for messages that invalidate the current request and cancel it in that case returning some error response.
2. We use LSP’s mechanism for cancelling requests and rely on the client’s knowledge of what it needs to invalidate.

Given all that, I am currently leaning towards keeping our existing request queue and only add cancellation based on LSP’s cancellation support to it. I think @ndmitchell wasn’t very keen on this in the past for reasons that I cannot remember so I’d like to get some feedback.

**Comment by neil-da** [comment]
Discussions concluded that:

* For goto definition and type on hover we should have a queue and process them in order. We should assume VS Code sends timely cancel requests so we can prune the queue. Text change events should be put in the queue as well. Whenever we run Shake we also fling in the ofInterest rule at low priority, so that it computes the requested information and general diagnostics. If we ever get to a point where we have nothing to do, we just run the general diagnostics.

* For code lens we treat it like things we can send to the server, by storing the last request id, and sending the data when we compute it.

**Comment by cocreature** [comment]
Request cancellation has now been implemented so I’m closing this. I don’t think it makes sense to do anything special for code lens requests now that they only depend on compiling the file rather than running the scenarios.
